src/btree/bt_sync_obsolete.c

```diff
@@ -291,3 +290,324 @@ __wt_sync_obsolete_cleanup(WT_SESSION_IMPL *session, WT_REF *parent)
+     *
+     * FIXME: Read internal pages from non-logged tables when the remove/truncate
+     * operation is performed using no timestamp.
+     */
+
+    if (addr.type == WT_ADDR_LEAF_NO ||
+      (addr.ta.newest_stop_durable_ts == WT_TS_NONE &&
+        (F_ISSET(S2C(session), WT_CONN_CKPT_CLEANUP_SKIP_INT) ||
+          !F_ISSET(S2BT(session), WT_BTREE_LOGGED)))) {
+        __wt_verbose_debug2(
+          session, WT_VERB_CHECKPOINT_CLEANUP, "%p: page walk skipped", (void *)ref);
+        WT_STAT_CONN_DATA_INCR(session, checkpoint_cleanup_pages_walk_skipped);
+        *skipp = true;
+    }
+    return (0);
+}
+
+/*
+ * __checkpoint_cleanup_walk_btree --
+ *     Check and perform checkpoint cleanup on the uri.
+ */
+static int
+__checkpoint_cleanup_walk_btree(WT_SESSION_IMPL *session, const char *uri)
+{
+    WT_BTREE *btree;
+    WT_DECL_RET;
+    WT_REF *ref;
+    uint32_t flags;
+
+    ref = NULL;
+    flags = WT_READ_NO_EVICT;
+
+    /*
+     * To reduce the impact of checkpoint cleanup on the running database, it operates only on the
+     * dhandles that are already opened.
+     */
+    WT_WITHOUT_DHANDLE(session,
+      WT_WITH_HANDLE_LIST_READ_LOCK(session, (ret = __wt_conn_dhandle_find(session, uri, NULL))));
+    if (ret == WT_NOTFOUND)
+        return (0);
+
+    /* Open a handle for processing. */
+    ret = __wt_session_get_dhandle(session, uri, NULL, NULL, 0);
+    if (ret != 0)
+        WT_RET_MSG(session, ret, "%s: unable to open handle%s", uri,
+          ret == EBUSY ? ", error indicates handle is unavailable due to concurrent use" : "");
+
+    btree = S2BT(session);
+    /* There is nothing to do on an empty tree. */
+    if (btree->root.page == NULL)
+        goto err;
+
+    /*
+     * FLCS pages cannot be discarded and must be rewritten as implicitly filling in missing chunks
+     * of FLCS namespace is problematic.
+     */
+    if (btree->type == BTREE_COL_FIX)
+        goto err;
+
+    /* Walk the tree. */
+    while ((ret = __wt_tree_walk_custom_skip(
+              session, &ref, __checkpoint_cleanup_page_skip, NULL, flags)) == 0 &&
+      ref != NULL) {
+        if (F_ISSET(ref, WT_REF_FLAG_INTERNAL))
+            WT_WITH_PAGE_INDEX(session, ret = __checkpoint_cleanup_obsolete_cleanup(session, ref));
+        WT_ERR(ret);
+
+        /* Check if we're quitting. */
+        if (!__checkpoint_cleanup_run_chk(session))
+            break;
+    }
+
+err:
+    /* On error, clear any left-over tree walk. */
+    WT_TRET(__wt_page_release(session, ref, flags));
+    WT_TRET(__wt_session_release_dhandle(session));
+    return (ret);
+}
+
+/*
+ * __checkpoint_cleanup_int --
+ *     Internal function to perform checkpoint cleanup of all eligible files.
+ */
+static int
+__checkpoint_cleanup_int(WT_SESSION_IMPL *session)
+{
+    WT_CURSOR *cursor;
+    WT_DECL_RET;
+    const char *uri;
+
+    WT_RET(__wt_metadata_cursor(session, &cursor));
+    while ((ret = cursor->next(cursor)) == 0) {
```

Comment:
Should we do a `search_near` or use `bounds` to satisfy the check below regarding `WT_BTREE_PREFIX`?
We want to have everything that starts with `file:` and `tiered:`, we could set a lower bound to `file:` and stop as soon as the uri does not start with `tiered`?